browser/modules/PlacesProvider.jsm

```diff
@@ -0,0 +1,258 @@
+   *         Order is ascending.
+   */
+  compareLinks: function LinkUtils_compareLinks(aLink1, aLink2) {
+    for (let prop of this._sortProperties) {
+      if (!(prop in aLink1) || !(prop in aLink2)) {
+        throw new Error("Comparable link missing required property: " + prop);
+      }
+    }
+    return aLink2.frecency - aLink1.frecency ||
+           aLink2.lastVisitDate - aLink1.lastVisitDate ||
+           aLink1.url.localeCompare(aLink2.url);
+  },
+};
+
+/**
+ * Singleton that serves as the default link provider for the grid. It queries
+ * the history to retrieve the most frequently visited sites.
+ */
+let Links = {
+  /**
+   * EventEmitter interface
+   */
+  eventEmitter: new EventEmitter(),
+  on: (...params) => this.eventEmitter.on(...params),
+  once: (...params) => this.eventEmitter.once(...params),
+  off: (...params) => this.eventEmitter.off(...params),
+
+  /**
+   * Set this to change the maximum number of links the provider will provide.
+   */
+  maxNumLinks: HISTORY_RESULTS_LIMIT,
+
+  /**
+   * A set of functions called by @mozilla.org/browser/nav-historyservice
+   * All history events are emitted from this object.
+   */
+  historyObserver: {
+    onDeleteURI: function PlacesProvider_onDeleteURI(aURI) {
+      // let observers remove sensetive data associated with deleted visit
+      Links.eventEmitter.emit("deleteURI", {
+        url: aURI.spec,
+      });
+    },
+
+    onClearHistory: function() {
+      Links.eventEmitter.emit("clearHistory");
+    },
+
+    onFrecencyChanged: function PlacesProvider_onFrecencyChanged(aURI,
+                           aNewFrecency, aGUID, aHidden, aLastVisitDate) { // jshint ignore:line
+      // The implementation of the query in getLinks excludes hidden and
+      // unvisited pages, so it's important to exclude them here, too.
+      if (!aHidden && aLastVisitDate) {
+        Links.eventEmitter.emit("linkChanged", {
+          url: aURI.spec,
+          frecency: aNewFrecency,
+          lastVisitDate: aLastVisitDate,
+          type: "history",
+        });
+      }
+    },
+
+    onManyFrecenciesChanged: function PlacesProvider_onManyFrecenciesChanged() {
+      Links.eventEmitter.emit("manyLinksChanged");
+    },
+
+    onTitleChanged: function PlacesProvider_onTitleChanged(aURI, aNewTitle) {
+      Links.eventEmitter.emit("linkChanged", {
+        url: aURI.spec,
+        title: aNewTitle
+      });
+    },
+
+    QueryInterface: XPCOMUtils.generateQI([Ci.nsINavHistoryObserver,
+                        Ci.nsISupportsWeakReference])
+  },
+
+  /**
+   * Must be called before the provider is used.
+   */
+  init: function PlacesProvider_init() {
+    PlacesUtils.history.addObserver(this.historyObserver, true);
+  },
+
+  /**
+   * Gets the current set of links delivered by this provider.
+   *
+   * @returns {Promise} Returns a promise with the array of links as payload.
+   */
+  getLinks: function PlacesProvider_getLinks() {
```

Comment:
Actually, ignore the above. I commented too early. It's fine to return a promise here.